hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java

```diff
@@ -0,0 +1,251 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.io.compress.zlib;
+
+import java.io.IOException;
+import java.util.zip.Checksum;
+import java.util.zip.Deflater;
+import java.util.zip.GZIPOutputStream;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.compress.Compressor;
+import org.apache.hadoop.io.compress.DoNotPool;
+import org.apache.hadoop.util.DataChecksum;
+
+/**
+ * A {@link Compressor} based on the popular gzip compressed file format.
+ * http://www.gzip.org/
+ */
+@DoNotPool
+public class BuiltInGzipCompressor implements Compressor {
+
+  /**
+   * Fixed ten-byte gzip header. See {@link GZIPOutputStream}'s source for
+   * details.
+   */
+  private static final byte[] GZIP_HEADER = new byte[]{
+          0x1f, (byte) 0x8b, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+  // The trailer will be overwritten based on crc and output size.
+  private byte[] GZIP_TRAILER = new byte[]{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+  private final int GZIP_HEADER_LEN = GZIP_HEADER.length;
+  private final int GZIP_TRAILER_LEN = GZIP_TRAILER.length;
+
+  private Deflater deflater;
+
+  private int headerOff = 0;
+  private int trailerOff = 0;
+
+  private int numBytesWritten = 0;
+
+  private int currentBufLen = 0;
+
+  private final Checksum crc = DataChecksum.newCrc32();
+
+  private BuiltInGzipDecompressor.GzipStateLabel state;
+
+  public BuiltInGzipCompressor(Configuration conf) { init(conf); }
+
+  @Override
+  public boolean finished() {
+    // Only if the trailer is also written, it is thought as finished.
+    return deflater.finished() && state == BuiltInGzipDecompressor.GzipStateLabel.FINISHED;
+  }
+
+  @Override
+  public boolean needsInput() {
+    if (state == BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
+      return deflater.needsInput();
+    }
+
+    return (state != BuiltInGzipDecompressor.GzipStateLabel.FINISHED);
+  }
+
+  @Override
+  public int compress(byte[] b, int off, int len) throws IOException {
+    int compressedBytesWritten = 0;
+
+    // If we are not within uncompressed data yet, output the header.
+    if (state != BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM &&
```

Comment:
Hmm, if we change to `HEADER_BASIC` here, we may also need to check if the state is `FINISHED`?

Reply:
I think `compress` is always used in loops like this:
```java
while (!compresser.needsInput()) {
  compresser.compress(..)
}
```
so if the state transitioned to `FINISHED`, we'd come out of the loop and ask for more input to compress.

Reply:
Hmm, okay. I set it to `HEADER_BASIC` and see if CI can pass.

Reply:
What I thought is, the caller might set input and compress until it doesn't need input. The state is in `FINISHED` and the caller might call set input and compress again. At the moment this check isn't effective to write the header.

Reply:
Hmm you are right. Should we change the state to `HEADER_BASIC` in `setInput`? it seems we should do so.

Reply:
Seems so. Let me update it.

Reply:
Hmm, oh, no, I think we cannot do it. `setInput` can be called multiple times before we reach finished status. If we set the state to `HEADER_BASIC`, it will re-output the header, but the current trailer is not output yet.

Reply:
Let me revert to original condition `INFLATE_STREAM` and `TRAILER_CRC`. It looks more reliable.

Reply:
Ah OK, so the compression can happen on multiple inputs. Then, I'm curious whether we should handle the `FINISHED` state too in this `if` clause. For instance, in the following situation:

```java
  @Override
  public void finish() throws IOException {
    if (!compressor.finished()) {
      compressor.finish();
      while (!compressor.finished()) {
        compress();
      }
    }
  }
```

the `CompressorStream` will first set the state to be `FINISHED` and then keep calling `compress` until it is finished.

Reply:
For `Compressor` here, `compress` requires a buffer to write compressed output to. So seems we cannot do like `CompressorStream`.

Once the caller calls `finish` on this `Compressor`, we only call `finish` on the deflator`. The caller then will call `finished` to verify if it reaches finished state. If not, it should call `compress` with buffer to get more compressed output.

Reply:
> So seems we cannot do like CompressorStream.

Hmm what do you mean here? sorry don't quite understand it.

> Once the caller calls finish on this Compressor, we only call finish on the deflator. The caller then will call finished to verify if it reaches finished state. If not, it should call compress with buffer to get more compressed output.

Yes. So if the input is large but the buffer in a `CompressorStream` is small, potentially it will need to call `compress` multiple times after the `finish()` is invoked, before it can reach to the `finished` status (i.e., `deflater.finished()` returns true).

Reply:
Oh, I mean `Compressor` doesn't have a `compress()` method. Once calling `compress`, the caller must provide a buffer for compressed output.

Calling `finish()` on this compressor won't set state to `FINISHED`. Only if we output the trailer, we set the state to `FINISHED`. In this state, the caller can set new input and call `compress(buf, off, len)` again. We will output new header and enter into another compressed output (section).

Reply:
I see. Yea I meant `CompressorStream` calls its own `compress` method which calls `compressor.compress` indirectly.

> Calling finish() on this compressor won't set state to FINISHED.

Oh sorry I was looking at an old version of this PR which still set it to `FINISHED` in `finish()`. Never mind.

So this looks good then :) Although I think the following:

> What I thought is, the caller might set input and compress until it doesn't need input. The state is in FINISHED and the caller might call set input and compress again. At the moment this check isn't effective to write the header.

Will also not happen? since when the state is finished, the caller should not call `setInput` before it calls `reset` on the compressor.

Reply:
Hmm, I'm not sure. :)

I looked at `finish()` at `Deflater`. It just sets a `finish` variable to true. But the variable is not used at all. So technically, I guess you still can call its `setInput` to set new input and `deflate` again?

Because of that, I take more conservative approach here in case of it.

Reply:
I think calling finish will set a flag to tell it flush the data when finished. 

Yea the current approach also looks good to me.